### ISSUE TITLE
Reduce boilerplate code with action type checks and casting

### ISSUE THREAD

**Issue by ranquild** [brianegan/dart_redux_epics#4]
I think we should use same approach as with redux itself:

```
_reducer = combineTypedReducers([
   new ReducerBinding<State, LoadSuccessAction>(_onLoadSuccess),
   new ReducerBinding<State, LoadFailureAction>(_onLoadFailure),
]);
```

we can create `combineTypedEpics` function with similar api:

```
_effect = combinedTypedEpics([
  new EpicBinding<State, LoadRequestAction(_onLoadRequest)
]);
```

where _onLoadRequest within epic can be defined like this:

```
Stream<Object> _onLoadRequest(Stream<LoadRequestAction> actions, EpicStore<State> store) {
   return actions.asyncMap((action) => api.getData(...).then((data) => new LoadSuccessAction(data)));
}
```

or even like this:
```
Stream<Object> _onLoadRequest(Stream<LoadRequestAction> actions, EpicStore<State> store) async* {
  await for(final action in actions) {
    final data = await api.getData(...);
    yield new LoadSuccessAction(data);
  }
}
```

Currently we have to add 
```
where((action) => action is LoadRequestAction)
.map((action) => action as LoadRequestAction)
```
to most epics.

We cannot use Rx Observables and ofType here because we have to merge stream with its "child" stream because of this https://github.com/brianegan/dart_redux_epics/issues/3.

**Comment by brianegan** [comment]
Makes sense, I was actually thinking the same thing, especially for folks who want to avoid Rx and use simple `async*` functions.

**Comment by brianegan** [comment]
Hey hey :)

I've got a fix for #3 and I've also got a fix to support `async*` functions (turned out to be a pretty cryptic bug with the Dart SDK, but I found a work-around).

I then gave this implementation a shot and got it working, but I'm actually not 100% sure I like it / whether it reduces boilerplate overall. Explanation following, please let me know your thoughts!

### With bindings

```dart
final epic = combinedTypedEpics([
  new EpicBinding<State, LoadRequestAction>(_onLoadRequest)
]);

Stream<dynamic> _onLoadRequest(
  Stream<LoadRequestAction> actions, 
  EpicStore<State> store,
) async* {
  await for(final action in actions) {
    final data = await api.getData(...);
    yield new LoadSuccessAction(data);
  }
}
```

### Without bindings

```dart
Stream<dynamic> _onLoadRequest(
  Stream<dynamic> actions, 
  EpicStore<State> store,
) async* {
  await for(final action in actions) {
    if (action is LoadRequestAction) {
      final data = await api.getData(...);
      yield new LoadSuccessAction(data);
    }
  }
}
```

The example without bindings:
  
  * Is actually 1 line shorter
  * Uses if statements, a basic language construct (reducing concepts the end user needs to learn)
  * Can be used independently without the needing to use the `combineTypedEpics` function
  * All logic is self-contained in 1 function, no need for an Epic + Binding for it to work

Whatcha think?

**Comment by iamarsey** [comment]
Wow, awesome improvements are coming, Great!
I would like to operate with typed bindings without additional 'if' statement. So, the approach with bindings looks more convenient and explicit.

**Comment by ranquild** [comment]
@brianegan I agree with @iamarsey, I would prefer typed epics, still, we can have both approaches. Just like with reducers - someone can actually do custom switch if this is required.

**Comment by brianegan** [comment]
Hey, I'm open to adding this but just want a bit more feedback...

  > I would like to operate with typed bindings without additional 'if' statement. So, the approach with bindings looks more convenient and explicit.

Could you explain your reasoning a bit more? I guess I don't see `ReducerBinding` as "more explicit" than an `if` statement. My thinking: `ReducerBinding` is a new abstraction and therefore *less* explicit.

  >  I would prefer typed epics because actual body of method is shorter

Agreed, but as you add more bindings, the difference in character count will continue to increase. If the goal of this is to reduce boilerplate, does this achieve that?

**Comment by ranquild** [comment]
@brianegan I understand your concern and I would say yes. Binding approach just looks more declarative to me. It doesn't look like actual code but more like a config file, which makes real code shorter. I would like to have both approaches (both with binding and without), if that possible, and we can give more feedback after month+ usage time.

**Comment by brianegan** [comment]
I getcha, since you can read down the list of bindings (even though they're kinda boilerplate), it makes it easier to scan what's happening.

Since you're using this in a bigger project than I am at the moment, I'll take your word for it and let's see how it goes.

I'll publish a new version of the library soon!

**Comment by brianegan** [comment]
Overall, I think we can actually simplify this a bit. What do you all think of this api:

```dart
combineEpics<dynamic>([
  new TypedEpic<dynamic, Request1>(fire1TypedEpic),
  new TypedEpic<dynamic, Request2>(fire2TypedEpic)
]);
```

  * No need for 2 types of combine functions
  * No need for "Bindings," that are converted to an epic. Just a `TypedEpic`!

**Comment by ranquild** [comment]
@brianegan I like it, looks great.

**Comment by brianegan** [comment]
Made me realize I can do this for Redux now as well, haha... Dang it! Wish I would have seen this pattern sooner :P

**Comment by brianegan** [comment]
Published as part of 0.7.0. Enjoy!

Note: I'll probably be updating this lib soon for Dart 2, which introduces some breaking changes.
